src/lib.js

```diff
@@ -0,0 +1,331 @@
+ * children by the first `bitWidth` bits of the key hash. In the next layer
+ * next `bitWidth` number of bits of the key hash are used to determine
+ * placement of it's children and so on. Each node in the tree will hold
+ * `2 * bitWidth` number of elements.
+ *
+ * For example when we insert an entry named `result` it will generate a
+ * folowing hash (with murmur3 32) `442064690` which in binary would be
+ *
+ * ```
+ * 11010010110010101111100110010
+ * ```
+ *
+ * Which will correspond to a following key path
+ *
+ * ```
+ * 11010/01011/00101/01111/10011/0010
+ * ```
+ *
+ * Whis in decimals would be
+ *
+ * ```ts
+ * 26/11/5/15/19/2'
+ * ```
+ *
+ * If we then insert `fish` it would produce following path
+ *
+ * ```ts
+ * 26/3/4/18/28/19/1
+ * ```
+ * @template [V=unknown]
+ * @template {string} [K=string]
+ * @template {API.Config} [C=API.Config<API.Uint32>]
+ * @param {Partial<C>} [options]
+ * @returns {API.PersistentHashMap<V, K, C>}
+ */
+export const empty = options => {
+  const config = configure(options)
+  return new PersistentHashMap(0, createBitmapIndexedNode(config, null), config)
+}
+
+/**
+ * @template {API.Config} [C=API.Config<API.Uint32>]
+ * @param {Partial<C>} config
+ */
+const configure = ({
+  bitWidth = 5,
+  BitField = configureBitField({ bitWidth }),
+  Path = bitWidth === 5
+    ? Uint32Path.configure({ bitWidth })
+    : Uint8ArrayPath.configure({ bitWidth }),
+} = {}) => /** @type {C} */ ({ bitWidth, BitField, Path })
+
+/**
+ * Creates HashMap from the provided entries.
+ *
+ * @template [V=unknown]
+ * @template {string} [K=string]
+ * @template {API.Config} [C=API.Config<API.Uint32>]
+ * @param {Iterable<[K, V]>} entries
+ * @param {Partial<C>} [options]
+ * @returns {API.PersistentHashMap<V, K, C>}
+ */
+export const from = (entries, options) => {
+  const node = /** @type {API.HashMapBuilder<V, K, C>} */ (builder(options))
+  for (const [key, value] of entries) {
+    node.set(key, value)
+  }
+
+  return node.build()
+}
+
+/**
+ * @template T
+ * @template {string} K
+ * @param {API.HAMT<T, K>} hamt
+ * @param {K} key
+ */
+export const has = (hamt, key) =>
+  Node.get(hamt.root, key, NOT_FOUND) !== NOT_FOUND
+
+/**
+ * @template T
+ * @template {string} K
+ * @template [U=undefined]
+ * @param {API.HAMT<T, K>} hamt
+ * @param {K} key
+ * @param {U} notFound
+ * @returns {T|U}
+ */
+export const get = (hamt, key, notFound = /** @type {U} */ (undefined)) =>
+  Node.get(hamt.root, key, notFound)
+
+/**
+ * @template {string} K
+ * @template T
+ * @template {API.Config} C
+ * @param {Partial<C>} [options]
+ * @returns {API.HashMapBuilder<T, K, C>}
+ */
+export const builder = options => {
+  const edit = {}
+  const config = configure(options)
+  return new HashMapBuilder(
+    edit,
+    0,
+    createBitmapIndexedNode(config, edit),
+    config
+  )
+}
+
+/**
+ * @template T
+ * @template {string} K
+ * @template {API.Config} C
+ * @implements {API.PersistentHashMap<T, K, C>}
```

Comment:
I believe `tsc` will not check this `@implements` clause. But there is a one-liner you can put in your constructor so tsc will ensure you implement what you want.
https://github.com/microsoft/TypeScript/issues/30156#issuecomment-952667829

Here's an example of how I used it in other repo https://github.com/web3-storage/web3.storage/blob/main/packages/api/src/utils/stripe.js#L43

Reply:
It does in fact check `implements` clause and complains if there is a mismatch. Why do you think it isn't ? Are there some cases when it does not ?